### DISCUSSION TITLE
TypeError: Cannot read property 'Title' of undefined

### DISCUSSION THREAD

**Discussion by jordan-miguel** [gatsbyjs/gatsby#30186] [Help]
## Issue

I can't successfully render my queried data inside of my component.

### Steps to reproduce

1. I test my graphQL query that will be used to pull in data from my CMS (_sanity_)
2. I attempt to render this in my component
3. I run `gatsby develop`
4. I can see `TypeError: Cannot read property 'Title' of undefined` error on my browser

I am using [these docs ](https://www.gatsbyjs.com/docs/recipes/querying-data/)as a reference.

### Code 

```
export const query = graphql`
  query {
    allSanityPosts {
      edges {
        node {
          Title
        }
      }
    }
  }
`;

const singlePost = ({ data }) => (
    <Layout>
        <Header />
        <div>
        <h1>{data.post.Title}</h1>
        </div>
    </Layout>
)

export default singlePost
```

### Error

![Screenshot 2021-03-10 at 23 27 50](https://user-images.githubusercontent.com/14114089/110711925-425a0380-81f8-11eb-9231-2baf09a2bf37.png)

**Comment by KyleAMathews** [comment]
Does the query work in graphiql?

**Comment by jordan-miguel** [comment]
Yes it does :) 

<img width="1211" alt="Screenshot 2021-03-11 at 00 02 54" src="https://user-images.githubusercontent.com/14114089/110715433-260c9580-81fd-11eb-97ec-881c786e537a.png">

**Comment by kimbaudi** [comment]
hey @jordan-miguel - i think you need to change from `data.post.Title` to `data.sanityPosts.Title`

---

Also, i think you can shorten your query from this:

```gql
export const query = graphql`
  query {
    allSanityPosts {
      edges {
        node {
          Title
        }
      }
    }
  }
`;
```

to this:

```gql
export const query = graphql`
  query {
    allSanityPosts {
      nodes {
        Title
      }
    }
  }
`;
```

**Reply by jordan-miguel** [reply]
hey @kimbaudi unfortunately Title is still an undefinded property.

**Reply by kimbaudi** [reply]
@jordan-miguel - can you try `data.allSanityPosts.edges.node` or `data.allSanityPosts.nodes` (if you changed the query like i suggested earlier)? Also try to add a `console.log(props)` to `singlePost` and see if `props.data` is returning a value.

```js
const singlePost = (props) => {
console.log(props)
const {data} = props
return (
    <Layout>
        <Header />
        <div>
        <h1>{data.post.Title}</h1>
        </div>
    </Layout>
)
}

```

I'm starting to suspect that you are receiving the error because you are using the `allSanityPosts` graphql query in a non-page component like a template, which wouldn't work. Where is the file located where you are using `allSanityPosts`? Is it under `src/pages` or `src/templates` or somewhere else?

**Reply by jordan-miguel** [reply]
@kimbaudi I'm able to successfully retrieve the data within the component with `data.sanityPosts.Title` & `data.sanityPosts.body`

The only issue I'm having is that my query retrives just a **_single post_**, regardless of whether or not I'm on `/blog/post-1` or `/blog/post-2` - (see code below).

I need to ensure that a **_specific post is fetched_**, I believe this can be achieved by adding a variable to my query, more specifically the `id` variable - based on [these docs](https://www.gatsbyjs.com/docs/how-to/querying-data/page-query/#how-to-add-query-variables-to-a-page-query). Does this make sense?

```
export const query = graphql`
  query {
    sanityPosts {
      Title
      body
    }
  }
`;

const singlePost = ({ data }) => (
    <Layout>
        <Header />
        <div>
        <h1>{data.sanityPosts.Title}</h1>
        <p>{data.sanityPosts.body}</p>
        </div>
    </Layout>
)

export default singlePost
```

**Reply by kimbaudi** [reply]
@jordan-miguel  - it is not enough for you to change `allSanityPosts` query to `sanityPosts` query. you also need to pass a parameter to your `sanityPosts` query so it can filter the correct post. you need to update `sanityPosts` to something like this:

```js
export const query = graphql`
  query($id: String!) {
    sanityPosts(id: { eq: $id }) {
      Title
      body
    }
  }
`;
```

**Reply by jordan-miguel** [reply]
@kimbaudi Now I'm seeing the error `TypeError: Cannot read property 'sanityPosts' of undefinded`

I'm not sure why, I still have this problem - Thank you for the help, and aplogies for the questions...I just don't understand why there are so many **undefined properties*.*
```
export const query = graphql`
  query($id: String!) {
    sanityPosts(id: { eq: $id }) {
      Title
      body
    }
  }
`;

const singlePost = ({ data }) => (
    <Layout>
        <Header />
        <div>
        <h1>{data.sanityPosts.Title}</h1>
        <p>{data.sanityPosts.body}</p>
        </div>
        <Footer />
    </Layout>
)

export default singlePost
```

**Reply by kimbaudi** [reply]
sorry I can't help anymore. I'm not entirely sure why you're getting the errors.
